nexus_motion_planner/src/motion_plan_cache.hpp

```diff
@@ -0,0 +1,138 @@
+#include <geometry_msgs/msg/pose.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <lifecycle_msgs/msg/state.hpp>
+
+// moveit modules
+#include <moveit/move_group_interface/move_group_interface.h>
+
+// NEXUS messages
+#include <nexus_endpoints.hpp>
+
+namespace nexus {
+namespace motion_planner {
+
+/**
+ * Cache manager for the Nexus motion planner.
+ *
+ * This manager facilitates cache management for MoveIt 2's `MoveGroupInterface`
+ * by using `warehouse_ros` to manage a database of executed motion plans, and
+ * how long they took to execute. This allows for the lookup and reuse of the
+ * best performing plans found so far.
+ *
+ * Relevant ROS Parameters:
+ *   - `warehouse_plugin`: What database to use
+ *   - `warehouse_host`: Where the database should be created
+ *   - `warehouse_port`: The port used for the database
+ *
+ * Motion plans are stored in the `move_group_plan_cache` database within the
+ * database file, with plans for each move group stored in a collection named
+ * after the relevant move group's name.
+ *
+ * For example, the "my_move_group" move group will have its cache stored in
+ * `move_group_plan_cache@my_move_group`
+ *
+ * Motion plans are keyed on:
+ *   - Plan Start: robot joint state
+ *   - Plan Goal (either of):
+ *     - Final pose (wrt. `planning_frame` (usually `base_link`))
+ *     - Final robot joint states
+ *
+ * Motion plans may be looked up with some tolerance.
+ */
+class MotionPlanCache
+{
+public:
+  // We explicitly need a Node::SharedPtr because warehouse_ros ONLY supports
+  // it...
+  explicit MotionPlanCache(const rclcpp::Node::SharedPtr& node);
+
+  void init(
+    const std::string& db_path = ":memory:",
+    uint32_t db_port = 0,
+    double exact_match_precision = 0.00001);
+
+  // TOP LEVEL OPS =============================================================
+  std::vector<
+    warehouse_ros::MessageWithMetadata<
+      moveit_msgs::msg::RobotTrajectory
+    >::ConstPtr
+  >
+  fetch_all_matching_plans(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double start_tolerance, double goal_tolerance, bool metadata_only = false);
+
+  warehouse_ros::MessageWithMetadata<
+    moveit_msgs::msg::RobotTrajectory
+  >::ConstPtr
+  fetch_best_matching_plan(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    double start_tolerance, double goal_tolerance, bool metadata_only = false);
+
+  bool put_plan(
+    const moveit::planning_interface::MoveGroupInterface& move_group,
+    const std::string& move_group_namespace,
+    const moveit_msgs::msg::MotionPlanRequest& plan_request,
+    const moveit_msgs::msg::RobotTrajectory& plan,
+    double execution_time_s,
+    double planning_time_s,
+    bool overwrite = true);
```

Comment:
In most cases, no. However:
- If `exact_match_precision` is set very high (relative to start and goal match tolerance), then overwriting will result in the cache being pretty sparse.
- Not overwriting results in the cache space being maximally dense, which maximizes the chances of a cache hit on lookup.
- Also it can be used to see historically how plans have evolved over time.

I think for all those reasons I'd like to keep a parameter here to disable overwriting, especially if we plan to upstream this cache class to MoveIt!

Note: Overwriting in the context of the cache means deleting **ALL** "worse" plans that have an "exact" match (according to the `exact_match_precision` amount, across all cache entry metadata fields), before inserting the new "better" entry.

Reply:
I think this behavior should be documented, iirc `put_plan` only inserts a new plan if it is better than existing plans. So `overwrite` has no effect if it is not the best plan.

Reply:
The insertion check only happens at the end, so technically speaking `overwrite` will delete all worse plans.

I decided to name it `overwrite` -> `delete_worse_plans` to be more clear.

f5713a1620c41a1f78adc6ffc899cac89f6bd64a in https://github.com/osrf/nexus/pull/26